source/extensions/common/async_files/async_file_handle.h

```diff
@@ -0,0 +1,81 @@
+  //
+  // If cancelled before the callback is called but after creating the file, unlinks the file.
+  virtual std::function<void()> createHardLink(absl::string_view filename,
+                                               std::function<void(absl::Status)> on_complete) PURE;
+
+  // Enqueues an action to close the currently open file.
+  // It is an error to use an AsyncFileContext after calling close.
+  // There must not already be an action queued for this handle.
+  // It is an error to destroy an AsyncFileHandle without calling close.
+  // It is an error to call close twice.
+  // Note that because an AsyncFileHandle is a shared_ptr, it is okay to
+  // call close during the handle's owner's destructor - that will queue the close
+  // event, which will keep the handle alive until after the close operation
+  // is completed. (But be careful that the on_complete callback doesn't require any
+  // context from the destroyed owner!)
+  virtual std::function<void()> close(std::function<void(absl::Status)> on_complete) PURE;
+
+  // Enqueues an action to read from the currently open file, at position offset, up to the number
+  // of bytes specified by length. The size of the buffer passed to on_complete informs you if less
+  // than the requested amount was read. It is an error to read on an AsyncFileContext that does not
+  // have a file open. There must not already be an action queued for this handle.
+  virtual std::function<void()>
+  read(off_t offset, size_t length,
+       std::function<void(absl::StatusOr<Buffer::InstancePtr>)> on_complete) PURE;
+
+  // Enqueues an action to write to the currently open file, at position offset, the bytes contained
+  // by contents. It is an error to call write on an AsyncFileContext that does not have a file
+  // open.
+  //
+  // This call consumes the 'contents' buffer immediately, by move, so it is safe to discard the
+  // buffer after the call, and is not safe to assume it still contains the same data.
+  //
+  // on_complete is called with the number of bytes written on success.
+  // There must not already be an action queued for this handle.
+  virtual std::function<void()> write(Buffer::Instance& contents, off_t offset,
+                                      std::function<void(absl::StatusOr<size_t>)> on_complete) PURE;
+
+  // Creates a new AsyncFileHandle referencing the same file.
+  // This is not asynchronous like the other file operations, as it is never a time-consuming
+  // operation. Note that a file handle duplicated in this way shares positioning and permissions
+  // with the original. Since AsyncFileContext functions are all position-explicit, this should not
+  // matter.
+  //
+  // The user should avoid using duplicate while a close, open or create operation is in flight on
+  // the source handle, to avoid provoking a race (the underlying file operation will provide an
+  // appropriate error in the event that a race goes the wrong way).
```

Comment:
Perhaps fail this via absl::StatusOr per above?

Reply:
Same, should no longer be racing because operations should all be sharing a thread/queue.